packages/expo-cli/src/commands/build/AndroidBuilder.ts

```diff
@@ -53,207 +56,43 @@ See https://docs.expo.io/distribution/building-standalone-apps/#2-configure-appj
 
-        // read the keystore
-        const keystoreData = await fs.readFile(keystorePath);
-
-        const credentials: AndroidCredentials.Keystore = {
-          keystore: keystoreData.toString('base64'),
-          keyAlias: keystoreAlias,
-          keystorePassword,
-          keyPassword,
-        };
-        await Credentials.updateCredentialsForPlatform(
-          ANDROID,
-          // @ts-ignore: Type '{ keystore: string; keystoreAlias: any; keystorePassword: any; keyPassword: any; }' has no properties in common with type 'Credentials'.
-          credentials,
-          [],
-          credentialMetadata
-        );
-      }
+    if (this.options.clearCredentials) {
+      await runCredentialsManager(ctx, new RemoveKeystore(experienceName));
     }
-  }
-
-  checkEnv(): boolean {
-    const allEnvSet =
-      !!this.options.keystorePath &&
-      !!this.options.keystoreAlias &&
-      !!process.env.EXPO_ANDROID_KEYSTORE_PASSWORD &&
-      !!process.env.EXPO_ANDROID_KEY_PASSWORD;
 
-    if (allEnvSet) {
-      return true;
-    }
-
-    // Check if user was trying to upload keystore incorretly and supply an helpful error message if so.
-    if (this.options.keystorePath || this.options.keystoreAlias) {
-      throw Error(
-        'When uploading your own keystore you must provide:\n' +
-          '\t--keystore-path /path/to/your/keystore.jks \n' +
-          '\t--keystore-alias PUT_KEYSTORE_ALIAS_HERE \n' +
-          'And set the enviroment variables:\n' +
-          '\tEXPO_ANDROID_KEYSTORE_PASSWORD\n' +
-          '\tEXPO_ANDROID_KEY_PASSWORD\n' +
-          'For details, see:\n' +
-          '\thttps://docs.expo.io/distribution/building-standalone-apps/#if-you-choose-to-build-for-android'
+    const paramKeystore = await getKeystoreFromParams(this.options);
+    if (paramKeystore) {
+      await useKeystore(ctx, experienceName, paramKeystore);
+    } else if (await keytoolCommandExists()) {
+      try {
+        await runCredentialsManager(ctx, new SetupAndroidKeystore(experienceName));
+      } catch (_) {
+        // error is logged inside credentials manager
+        // keystore will be generated by worker
+      }
+    } else {
+      log.warn(
+        'Program keytool was not found in PATH, new Keystore will be generated on Expo servers'
       );
     }
-    return false;
-  }
-
-  async collectAndValidateCredentialsFromCI(
-    credentialMetadata: Credentials.CredentialMetadata
-  ): Promise<void> {
-    const credentials: any = {
-      keystore: (await fs.readFile(this.options.keystorePath!)).toString('base64'),
-      keystoreAlias: this.options.keystoreAlias,
-      keystorePassword: process.env.EXPO_ANDROID_KEYSTORE_PASSWORD,
-      keyPassword: process.env.EXPO_ANDROID_KEY_PASSWORD,
-    };
-    await Credentials.updateCredentialsForPlatform(ANDROID, credentials, [], credentialMetadata);
   }
+}
 
-  platform(): Platform {
-    return ANDROID;
+async function keytoolCommandExists(): Promise<boolean> {
```

Comment:
the only issue with compatibility I found is that on some version store type defaulted to PKCS instead JKS, but it was fixed some time ago. If local generation of keystore fails we are not stoping build command.

Reply:
> If local generation of keystore fails we are not stoping build command.

I removed that behaviour because it complicated code significantly, but I don't think this will be a problem. It would be very weird case for keytool to exist in PATH but fail to generate keystore